tests/integ/test_temp_table_cleanup.py

```diff
@@ -0,0 +1,63 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import copy
+import gc
+
+import pytest
+
+from snowflake.snowpark.functions import col
+
+pytestmark = [
+    pytest.mark.skipif(
+        "config.getoption('local_testing_mode', default=False)",
+        reason="Temp table cleanup is not supported in Local Testing",
+    ),
+]
+
+
+def test_reference_count_map_basic(session):
+    df1 = session.create_dataframe([[1, 2], [3, 4]], schema=["a", "b"]).cache_result()
+    table_name = df1.table_name
+    df1.collect()
+    assert session._temp_table_ref_count_map[table_name] == 1
+    df2 = df1.select("*").filter(col("a") == 1)
+    df2.collect()
+    assert session._temp_table_ref_count_map[table_name] == 1
+    df3 = df1.union_all(df2)
+    df3.collect()
+    assert session._temp_table_ref_count_map[table_name] == 1
+    del df1
```

Comment:
Yes, if a dataframe is used inside a function but not returned, it will be garbage collected upon completing calling that function. Added a test for that